### PR TITLE
[Withdrawn] restore A5 full-Acc and Vec layout correctness

### PR DESCRIPTION
**Pull request by lwDavid**
## Summary

- preserve a distinct result handle for full-Acc `tile.matmul_acc` values that reuse the accumulator's physical L0C address
- keep true, unmaterialized Acc subviews in place so MAD still updates their parent window
- prevent A5 Vec ND and NZ representation families from sharing one physical address in both the legacy and DSA-RP memory planners
- retain the existing A2/A3 cross-layout reuse policy and document the target-specific constraint
- make the A5 policy work for standalone pass invocation even when no `PassContext` is active

## Root cause

Two independent compiler regressions prevented the DeepSeek V4 Flash path from producing stable end-to-end tokens:

1. A full accumulator was lowered with the same MLIR handle as its loop-carried `matmul_acc` input. This changed PTOAS full-accumulator loop lowering and could stall long-running kernels. Only a real `pto.subview` needs the in-place handle introduced for subview correctness.
2. On A5, lifetime-disjoint Vec values using ND and NZ representations could reuse one address. Mixed kernels can then read representation-incompatible contents and silently corrupt decode values.

The layout restriction is queried through the configured backend handler and applies only when the PTO target architecture is `a5`. When a pass runs without an active `PassContext`, it falls back to the globally configured backend instead of silently disabling the A5 gate.

## Validation

- [x] Current worktree build at `be24b058` is clean and up to date on `origin/main` `bd3a477d`
- [x] Standalone no-`PassContext` regression verifies A5 separates ND/NZ while A2/A3 retains reuse
- [x] Affected test set: 169 passed
- [x] Full isolated unit suite: 10331 passed, 3 skipped, 2 xfailed
- [x] GitHub CI: 14/14 jobs passed, including unit, codegen, A5 mixed-kernel, model, direct/distributed system, and A5 simulator tests
- [x] `git diff --check`, Ruff check/format, pre-commit, clang-tidy, and documentation checks

The final-head local tests loaded both the Python package and native extension from the current worktree staging while retaining the shared dependency environment. The primary PyPTO editable finder was explicitly absent. The current native extension SHA-256 is `50d0b73b65249bc567dea52b83e65c0975b070d31e2d54678b1ed70f219c966d`.

The patch-equivalent pre-rebase stack (`6a3a8b25`) was also validated with the companion repository fixes on 8 devices using real DeepSeek V4 Flash EP8/TP2 weights: prefill 16 plus 32 decode steps completed, all 33 token markers were finite, rank-consistent, and greedy, with steady-state TPOT 213.861 ms/token. That end-to-end result depends on all companion PRs below; it is not a PyPTO-only validation.

## Companion PRs

- pypto-lib: https://github.com/hw-native-sys/pypto-lib/pull/1019
- PTO-ISA: https://gitcode.com/CANN/pto-isa/merge_requests/1525

## Risk

- A5 Vec peak memory may increase where ND and NZ lifetimes previously shared storage; same-representation reuse remains enabled.
- DSA-RP materializes cross-layout hard relations only for disjoint A5 Vec intervals.
- The `matmul_acc` codegen change is limited to full Acc values; actual unmaterialized subviews keep the in-place path.

### CODE REVIEW

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 1]
```diff
@@ -2165,13 +2181,15 @@ std::map<VarPtr, VarPtr> IdentifyReuseOpportunities(
            pipeline_load_tiles.count(b.variable.get()) != 0;
   };
 
+  const bool enforce_vec_nd_nz_layout = IsA5Target();
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply src/ir/transforms/memory_reuse_pass.cpp:2186]: Fixed in be24b058: `IsA5Target()` now falls back to the configured `BackendConfig` handler when no `PassContext` is active. The regression runs the public standalone pipeline on a fresh worker thread, explicitly asserts `PassContext.current() is None`, and verifies A5 separates ND/NZ while A2/A3 still reuses. The complete affected set passes (169 tests).

**Comment by lwDavid** [comment]
Validation update for `be24b058`:

- Fixed the no-`PassContext` A5 target lookup and added a standalone-path regression that explicitly verifies the worker has no active context.
- Updated the split-K boundary test for distinct full-Acc result handles while still pinning both the tail result valid shape and the seed `set_validshape` pair.
- Focused affected suite: **169 passed**.
- Full isolated unit suite on the current branch build: **10331 passed, 3 skipped, 2 xfailed**.
- GitHub CI: **14/14 jobs passed**; Docs build and CodeRabbit also passed (`deploy` is normally skipped).

The earlier local provenance issue was excluded: validation loaded both `pypto` and `pypto_core` from the current branch staging and verified that the primary editable finder was absent.

**Comment by lwDavid** [comment]
Evidence audit update: this PR is being returned to Draft while its two claimed fixes are revalidated independently.

The Vec ND/NZ change is withdrawn. The minimal case demonstrated an allocation-policy difference, not incorrect code generation. A serialized, same-device Ascend 950 gate-off/gate-on A/B changed the targeted addresses but both configurations passed all five production golden checks. The historical #2100/#2451 failures involved an actual incompatible consumer or direct move alias and do not establish that unrelated, lifetime-disjoint values cannot reuse raw UB storage. Issue #2517 has therefore been closed as not planned.

The full-Acc result-handle change is still under an isolated baseline-vs-patch audit. It will remain only if unmodified current lowering can be made to fail reproducibly and changing only that lowering restores the numerical/runtime result. Otherwise the whole PR will be closed rather than proposing an unproven code-generation restriction.

**Comment by lwDavid** [comment]
Closing this PR after a controlled, single-variable evidence audit. Neither proposed compiler change has a reproducible baseline failure that becomes a pass when only that change is applied, so these changes should not be proposed as bug fixes.

### Vec ND/NZ reuse policy

The CPU reproducer proved only that the allocator can reuse one raw Vec allocation for unrelated, lifetime-disjoint ND- and NZ-view values. It did not reproduce incorrect numerical behavior.

A serialized same-device Ascend 950 production A/B changed only this policy and confirmed that the intended address change occurred:

- gate off: the targeted ND/NZ pairs reused addresses (`49408/49408`, `16384/16384`, and `24576/24576`);
- gate on: the same pairs were separated (`57344/16384`, `41344/16384`, and `24576/41344`).

Nevertheless, both builds passed all five repository golden checks: `inner_compress_state`, `idx_kv_cache`, `idx_kv_scale`, `score`, and `topk_idxs` (`task_20260825_152106_26201176305`, `gate_off_rc=0`, `gate_on_rc=0`). Historical failures in #2100 and #2451 involved a direct layout-conversion alias or an actually incompatible consumer; they do not prove that unrelated, lifetime-disjoint values cannot reuse raw UB storage. Issue #2517 was therefore closed as invalid/not planned.

### Full-Acc result handle

The claimed same-handle stall was also not reproduced:

- the repository A5 full-Acc golden tests passed both without and with the change (2/2 in each build);
- a 255-iteration `64x16384x64` full-Acc stress test passed exact, zero-tolerance golden comparison without and with the change and did not stall;
- the production-like stage-2 nested-pipeline regression passed without and with the change;
- PTOAS 0.57 and PTO-ISA explicitly support same-handle `tmatmul.acc` / `TMATMUL_ACC`, and the emitted MAD remains same-handle even after this patch; the patch only introduces another same-address handle;
- a fresh strict real-weight DeepSeek V4 Flash EP8/TP2 run built from `be24b058` with only commit `1956f3db` reversed passed prefill plus two decode steps on 8 devices: tokens `11111`, `16`, and `983`, all active logits finite, all 8 ranks identical, and all three `greedy_match=True` (`task_20260825_162638_18600027401`, exit 0; run-log SHA-256 `6e24a08dab146ae2c03eab319b5f131b93222777931f4327a7189f6ed193746a`). Its 9/384/365 binary aggregates were unchanged before/after and identical to the patched artifacts.

The earlier handle-only end-to-end run still exhibited the same stall, so the eventual multi-repository recovery cannot be attributed to this PyPTO handle change.

The branch is retained only as audit history. A replacement PR should be opened only after a locked-stack, baseline-fail / single-patch-pass reproducer demonstrates a real compiler defect.
